api/api/V1/Processing.ts

```diff
@@ -352,10 +367,193 @@ export default function (app: Application, baseUrl: string) {
+      if (!recording) {
+        return next(
+          new AuthorizationError(
+            `Could not find a Recording with an id of '${request.params.id}'`,
+          ),
+        );
+      }
+      const data = response.locals.data;
+      const trackIds: number[] = [];
+      const tracks = [];
+      for (const trackData of data) {
+        const isMasked = await isTrackMasked(
+          recording,
+          trackData,
+          request.body.algorithmId,
+        );
+        if (isMasked) {
+          continue;
+        }
+        const track = prepareTrackToSave(
+          recording,
+          trackData,
+          request.body.algorithmId,
+        );
+        tracks.push(track);
+      }
+      const modelTracks = await Track.bulkCreate(tracks);
+
+      const trackTags = [];
+      const trackTagData = [];
+      for (let i = 0; i < modelTracks.length; i++) {
+        const track = tracks[i];
+        const trackData = data[i];
+        trackIds.push(modelTracks[i].id);
+
+        for (const pred of trackData.predictions) {
+          const modelName = pred.name;
+          const used = modelName === AI_MASTER;
+          let confidence = pred.confidence;
+          // confidence will always be over 50 if it was already put in the 0-100 range
+          if (confidence < 1) {
+            confidence = Math.round(100 * confidence);
+          }
+          const predData = pred as TrackTagData;
+          if (!pred.confident) {
+            predData.raw_tag = pred.tag;
+            pred.tag = "unidentified";
+          }
+
+          const tag = {
+            TrackId: modelTracks[i].id,
+            what: pred.tag,
+            confidence: pred.confidence,
+            automatic: true,
+            model: modelName,
+            UserId: null,
+            used,
+          } as TrackTag;
+          trackTags.push(tag);
+
+          trackTagData.push(predData);
+        }
+
+        delete trackData.predictions;
+        await Track.saveTrackData(track.id, trackData);
+      }
+
+      const modelTrackTags = await TrackTag.bulkCreate(trackTags);
+      for (let i = 0; i < modelTrackTags.length; i++) {
+        const modelTrackTag = modelTrackTags[i];
+        if (modelTrackTag.model) {
+          // Save the additional Track metadata to object storage
+          await Track.saveTrackTagData(modelTrackTag.id, trackTagData[i]);
+        }
+      }
+
+      return successResponse(response, "Tracks added.", { trackIds });
+    },
+  );
+
+  const isTrackMasked = async (
+    recording: Recording,
+    trackData: MinimalTrackRequestData,
+    algorithmId: number,
+  ): Promise<boolean> => {
+    const deviceId = recording.DeviceId;
+    const groupId = recording.GroupId;
+    const atTime = recording.recordingDateTime;
+    if (recording.type === RecordingType.ThermalRaw) {
+      const positions = trackData && trackData.positions;
+      if (positions) {
+        return trackIsMasked(deviceId, groupId, atTime, positions);
```

Comment:
`trackIsMasked` is an async function because it needs to lookup the `DeviceHistory` to see if there was masking.  If we're adding lots of tracks for a single recording/device at once, we're duplicating this effort unnecessarily.  It would be better to refactor this to pull out the mask region once, and then test against it for every track.